### PR TITLE
Mongo persistence

### PR DESCRIPTION
**Pull request by rosesyrett**

### CODE REVIEW

File: src/diffcalc_API/routes/constraints.py [hunk 1]
```diff
@@ -1,52 +1,73 @@
...
-async def get_constraints(name: str, store: HklCalcStore = Depends(get_store)):
-    content = await service.get_constraints(name, store)
-
+async def get_constraints(
+    name: str,
+    store: HklCalcStore = Depends(get_store),
+    collection: Optional[str] = Query(default=None, example="B07"),
+):
+    content = await service.get_constraints(name, store, collection)
     return Response(content=content, media_type="application/text")
```
Comment (callumforrester) [inline src/diffcalc_API/routes/constraints.py:18]: I like the inclusion of collections, two things to consider:

1. Should they take this form or does it make more sense to group them in the URL? For example:
```python
@router.get("/{collection}/{name}")
def get_constraints(collection: str, name: str, store: HklCalcStore = Depends(get_store)):
    ...
```

2. Should they go in a separate PR?

@garryod and @RAYemelyanova, thoughts?

File: src/diffcalc_API/stores/mongo.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    async def create(self, name: str, collection: Optional[str]) -> None:
+        coll: Collection = database[collection if collection else "default"]
+
+        attempting_to_overwrite(name, coll)
+        ubcalc = UBCalculation(name=name)
+        constraints = Constraints()
+        hkl = HklCalculation(ubcalc, constraints)
+
+        coll.insert_one(hkl.asdict)
```
Comment (callumforrester) [inline src/diffcalc_API/stores/mongo.py]: This is a blocking I/O call inside an `async` method, the rest are the same. It's unlikely to scale well. There is an async-friendly library for accessing mongo: https://github.com/mongodb/motor, the API looks similar so not many code changes needed.

We should either use that or stop supporting async. Again, I welcome opinions.

**Review by callumforrester — CHANGES_REQUESTED** [review]
Looks good, a few minor tweaks and possibly some splitting needed, see comments.

Also, doesn't currently work out-of-box as you need to manually install the master branch of diffcalc-core. Do we know if a release is likely in the near future?

File: src/diffcalc_API/routes/constraints.py (hunk 1 above)
Comment (garryod) [reply src/diffcalc_API/routes/constraints.py:18]: I reckon they're better suited in the URL as a user is likely to work within a given `collection`, but it's possible I don't have the correct mental model of how the service will be used

File: src/diffcalc_API/stores/mongo.py (hunk 2 above)
Comment (garryod) [reply src/diffcalc_API/stores/mongo.py]: Seems like the right call to me, other option being to wrap the synchronous calls in `asyncio` `Task`s

File: src/diffcalc_API/routes/constraints.py (hunk 1 above)
Comment (rosesyrett) [reply src/diffcalc_API/routes/constraints.py:18]: Ideally, yes these would go into a seperate PR. However I've already made changes in this branch, I feel like just going ahead with implementing it how I've done it, and then making another PR if I want to change it to this method.

The reason I thought a query was better for collection, is so that there is a default collection for users who dont want their crystals associated with a specific beamline. However, this sounds messy and I don't know how practical a use case it is.

File: src/diffcalc_API/stores/mongo.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    DiffcalcAPIException,
+    ErrorCodes,
+)
+from diffcalc_API.stores.protocol import HklCalcStore
+
+
+class Codes(ErrorCodes):
+    attempting_to_overwrite = 405
+    document_found = 404
+    nothing_to_delete = 404
```
Comment (garryod) [inline src/diffcalc_API/stores/mongo.py]: This could be an `Enum` or even `IntEnum`, which would provide the functionality of `ErrorCodes`

File: src/diffcalc_API/stores/mongo.py (hunk 2 above)
Comment (rosesyrett) [reply src/diffcalc_API/stores/mongo.py]: I agree with this: Moving to motor makes sense, to me it makes sense to use asyncio here as it's a call to a database.

**Comment by rosesyrett** [comment]
Just a comment - diffcalc-core has now been released with my changes so I no longer need to pull from the branch manually. I'll update this now.

Edit: my changes have been merged into master but a new version on pypi hasn't been released yet. So, not sure - am speaking with Irakli about this now.

File: src/diffcalc_API/stores/mongo.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    existing_document = collection.find_one({"ubcalc.name": name})
+    if existing_document:
+        message = (
+            f"Document already exists for crystal {name}!"
+            f"\nEither delete via DELETE request to this URL "
+            f"or change the existing properties. "
+        )
+        raise DiffcalcAPIException(
+            status_code=Codes.attempting_to_overwrite, detail=message
+        )
```
Comment (garryod) [inline src/diffcalc_API/stores/mongo.py]: `error_if_overwrite` would be a more descriptive name as the side effect of this function is not captured by `attempting_to_overwrite`

An alternative structure may be to create a `OverwriteError` which subclasses DiffcalcAPIException and formats an appropriate message, then check if an overwrite is being performed in the `create` method itself

File: src/diffcalc_API/stores/mongo.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+
+def document_found(name: str, retrieved_document: Optional[Dict[str, Any]]):
+    if not retrieved_document:
+        message = f"Document for crystal {name} not found! Cannot load."
+        raise DiffcalcAPIException(status_code=Codes.document_found, detail=message)
```
Comment (garryod) [inline src/diffcalc_API/stores/mongo.py]: Similar to above, perhaps `error_if_document_not_found`

File: src/diffcalc_API/stores/mongo.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+
+def nothing_to_delete(name: str, result: DeleteResult):
+    if result.deleted_count == 0:
+        message = f"Document for crystal {name} not found! Cannot delete"
+        raise DiffcalcAPIException(status_code=Codes.nothing_to_delete, detail=message)
```
Comment (garryod) [inline src/diffcalc_API/stores/mongo.py]: Similar to above, perhaps `error_if_nothing_to_delete`

File: src/diffcalc_API/stores/mongo.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+    async def load(self, name: str, collection: Optional[str]) -> HklCalculation:
+        coll: Collection = database[collection if collection else "default"]
+        hkl_json: Optional[Dict[str, Any]] = coll.find_one({"ubcalc.name": name})
+        document_found(name, hkl_json)
+        return HklCalculation.fromdict(hkl_json)
+
+
+def get_store() -> HklCalcStore:
+    return MongoHklCalcStore()
```
Comment (garryod) [inline src/diffcalc_API/stores/mongo.py]: A nicer pattern may be to have a dictionary of registered stores and a registration decorator in `protocol.py`, e.g.:

```python
STORES: dict[str, HklCalcStore] = dict()

def register_store(name: str):
    def do_registration(cls: HklCalcStore):
        STORES[name] = cls
    
    return do_registration
```

This would allow you to register stores like so:
```python

@register_store("mongo")
class MongoHklCalcStore:
    ...
```

And access them like so:

```python
store = STORES["mongo"]
```

The nice thing about this being that this registration could be performed by another python package, and the store to use can easily be passed via the CLI

**Review by garryod — CHANGES_REQUESTED** [review]

File: src/diffcalc_API/routes/constraints.py (hunk 1 above)
Comment (rosesyrett) [reply src/diffcalc_API/routes/constraints.py:18]: I think I'll make a separate PR with this. I'll just keep the collection as a query for now.

File: src/diffcalc_API/stores/mongo.py (hunk 3 above)
Comment (rosesyrett) [reply src/diffcalc_API/stores/mongo.py]: The reason it inherits from ErrorCodes is so it can use ErrorCodes.all_codes() method. How would I keep this functionality if I were to do this?

File: src/diffcalc_API/stores/mongo.py (hunk 4 above)
Comment (rosesyrett) [reply src/diffcalc_API/stores/mongo.py]: I like the sound of the second option. But I think it's best to do that in a separate PR, as my 'pickling.py' file in this folder does the same thing. Noted.

File: src/diffcalc_API/stores/mongo.py (hunk 7 above)
Comment (rosesyrett) [reply src/diffcalc_API/stores/mongo.py]: Wow, this is brilliant. Thanks, I'll update that. I was thinking there's got to be a better way of doing it...

Edit: have just realised I'll have to tweak what you've done slightly. These stores are used as fastAPI dependencies, so they must be no argument callables (i.e. I actually called get_store() before). So, STORES should be a dict of str, callable which returns HklCalcStore. But I think I can do this.

File: src/diffcalc_API/stores/mongo.py (hunk 4 above)
Comment (callumforrester) [reply src/diffcalc_API/stores/mongo.py]: +1 to the subclassing idea

File: setup.cfg [hunk 8]
```diff
@@ -27,6 +27,8 @@ install_requires =
     diffcalc-core
     fastapi
     uvicorn
+    pymongo
```
Comment (callumforrester) [inline setup.cfg:30]: Can remove the pymongo dependency now we have motor

File: src/diffcalc_API/stores/mongo.py (hunk 3 above)
Comment (garryod) [reply src/diffcalc_API/stores/mongo.py]: You'd probably still want an `ErrorCodes` class like so:

```python
class ErrorCodes(IntEnum):
    @classmethod
    @property
    def values(cls) -> set[str]:
        return {val.value for val in cls}
```

Allowing you to get a set of values with `ErrorCodes.values`

File: setup.cfg (hunk 8 above)
Comment (rosesyrett) [reply setup.cfg:30]: Ah, actually I use some of the types from pymongo because motor uses them also. So I've kept it in
